examples/java/src/main/java/org/apache/beam/examples/complete/TwitterStreamGenerator/TwitterIO.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * License); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.examples.complete.TwitterStreamGenerator;
+
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import twitter4j.TwitterStream;
+import twitter4j.*;
+import twitter4j.conf.ConfigurationBuilder;
+
+import java.io.Serializable;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+public class TwitterIO {
+
+    /**
+     * Implementation of read methods.
+     */
+    public static class Read extends PTransform<PBegin, PCollection<Read.TwitterConnection>> {
+
+        static BlockingQueue<Status> queue;
+        static String key;
+        static String secret;
+        static String token;
+        static String tokenSecret;
+
+        public Read(Builder builder) {
+            key = builder.key;
+            secret = builder.secret;
+            token = builder.token;
+            tokenSecret = builder.tokenSecret;
+            queue = new LinkedBlockingQueue<>();
+        }
+
+        @Override
+        public PCollection<TwitterConnection> expand(PBegin input) throws IllegalArgumentException {
+            if (key == null || secret == null || token == null || tokenSecret == null) {
+                throw new IllegalArgumentException("Please provide key, secret, token and token secret");
+            }
+
+            return input.apply(Create.of(TwitterConnection.getInstance()));
```

Comment:
here we also want to apply the ReadFromTwitterDoFn, right? And our return would be the contents of a tweet (e.g. `twitter4j.Status` or something like that)

Reply:
I didn't like this way of doing it, I don't wish to hide complexity from the user like this. I want to show how exactly the data is being read from the pipeline. I have re-organized this after reading the KafkaIO. My Idea is User giving us connection parameters, I converting that connection parameter to Twitter Config, This config then goes to `ReadFromTwitterDoFn` and a singleton Twitter connection is formed to fetch data. My Only concern here is that is ReadFromTwitterDoFn has a parallelism >1, that may create multiple instances of `Twitter connection` from different machines, even if it's a defined singleton. Does that make sense? Will it be better to putting ReadFromTwitterDoFn logic in `TwitterIO.java`? I need to think more on this

Reply:
that's reasonable. We have some transforms that do something like this. For example, in TextIO, we have the `ReadFiles` transform, which receives a PCollection of files, and returns the contents of these files.

See: https://beam.apache.org/releases/javadoc/2.29.0/org/apache/beam/sdk/io/TextIO.html#readFiles--

You could have two of them.

First the simple one that takes a connection as a parameter at pipeline construction:
```
pipeline.apply(TwitterIO.read()
  .withConnection(...))
```

and the second one, which takes a PCollection of Connections

```
pipeline.apply(Create.of(TwitterConnection.create(...), TwitterConnection.create(...)))
            .apply(TwitterIO.readAll())
```

WDYT?

Reply:
you are correct that each element would force a new connection to Twitter to be created

Reply:
Ok done, I think you are right, I think I still can work on the split logic, currently, the split is not actually happening.